ops/model.py

```diff
@@ -921,8 +928,34 @@ def _load(self) -> '_RelationDataContent_Raw':
             # Dead relations tell no tales (and have no data).
             return {}
 
-    def _is_mutable(self):
+    def _is_readable(self):
+        """Return if the data content can be read."""
+        if not self._strict_access_control:
+            return True
+
+        # Only remote units (and the leader unit) can read *this* app databag.
+
+        # is this an app databag?
+        if self._is_app:
+            # is this a LOCAL app databag?
+            if self._backend.app_name == self._entity.name:
+                # are we minions?
+                if not self._backend.is_leader():
+                    # type guard; we should not be accessing relation data
+                    # if the remote app does not exist.
+                    app = self.relation.app
+                    assert app is not None
+
+                    # is this a non-peer relation?
+                    if app.name != self._entity.name:
+                        return False
```

Comment:
Maybe check for the opposite side of this - e.g. check for `if i_am_remote or i_am_leader: return True` - seems like it could clean up and simplify this logic.

Reply:
I iterated four times on that code and couldn't find any other way to make it clear to myself.
But now that you say it it makes sense. 

*feels old in Python*

Reply:
gave it a shot, it's like four times as long now but less nesty. Added a bunch of comments because IMHO it's really not that easy